src/main/java/org/truffleruby/RubyContext.java

```diff
@@ -257,7 +266,97 @@ public void initialize() {
             sharedObjects.startSharing();
         }
 
+        if (isPreInitializing()) {
+            // Cannot save the FileDescriptor in the image
+            random = null;
+            // Cannot save native functions in the image
+            truffleNFIPlatform = null;
+            // Cannot save the root Java Thread instance in the image
+            threadManager.resetMainThread();
+        } else {
+            initialized = true;
+        }
+
+        this.preInitializing = false;
+    }
+
+    /** Re-initialize parts of the RubyContext depending on the current process */
+    private void reInitialize() {
+        if (random != null) {
+            throw new UnsupportedOperationException("re-initialization was already performed");
+        }
+
+        this.random = new SecureRandom();
+
+        this.truffleNFIPlatform = createNativePlatform();
+        coreLibrary.initializeDefaultEncodings(truffleNFIPlatform, nativeConfiguration);
+
+        threadManager.restartMainThread(Thread.currentThread());
+        threadManager.initialize(truffleNFIPlatform, nativeConfiguration);
+    }
```

Comment:
What's the difference between re-initializing and patching?

Reply:
We need to `reInitialize()` (restore the Java Thread, etc) also in finalizeContext (see the note there).
It's a part of patching.